ch390/src/esp_eth_mac_ch390.c

```diff
@@ -0,0 +1,908 @@
+{
+    emac_ch390_t *emac = __containerof(mac, emac_ch390_t, parent);
+    emac->flow_ctrl_enabled = enable;
+    return ESP_OK;
+}
+
+static esp_err_t emac_ch390_set_peer_pause_ability(esp_eth_mac_t *mac, uint32_t ability)
+{
+    emac_ch390_t *emac = __containerof(mac, emac_ch390_t, parent);
+    // we want to enable flow control, and peer does support pause function
+    // then configure the MAC layer to enable flow control feature
+    if (emac->flow_ctrl_enabled && ability) {
+        ch390_enable_flow_ctrl(emac, true);
+    }
+
+    else {
+        ch390_enable_flow_ctrl(emac, false);
+        ESP_LOGD(TAG, "Flow control not enabled for the link");
+    }
+    return ESP_OK;
+}
+
+static esp_err_t emac_ch390_transmit(esp_eth_mac_t *mac, uint8_t *buf, uint32_t length)
+{
+    esp_err_t ret = ESP_OK;
+    emac_ch390_t *emac = __containerof(mac, emac_ch390_t, parent);
+    uint8_t tcr = 0;
+
+    ESP_GOTO_ON_FALSE(length <= ETH_MAX_PACKET_SIZE, ESP_ERR_INVALID_ARG, err,
+                      TAG, "frame size is too big (actual %lu, maximum %u)",
+                      length, ETH_MAX_PACKET_SIZE);
+
+    /* copy data to tx memory */
+    ESP_GOTO_ON_ERROR(ch390_io_memory_write(emac, buf, length), err, TAG,
+                      "write memory failed");
+
+    /* Check if last transmit complete */
+    int64_t wait_time = esp_timer_get_time();
+    do {
+        ESP_GOTO_ON_ERROR(ch390_io_register_read(emac, CH390_TCR, &tcr), err, TAG,
+                          "read TCR failed");
+    } while ((tcr & TCR_TXREQ) && ((esp_timer_get_time() - wait_time) < CH390_MAC_TX_WAIT_TIMEOUT_US));
+
+    if (tcr & TCR_TXREQ) {
+        ESP_LOGE(TAG, "last transmit still in progress, cannot send.");
+        return ESP_ERR_INVALID_STATE;
+    }
+
+    /* set tx length */
+    ESP_GOTO_ON_ERROR(ch390_io_register_write(emac, CH390_TXPLL, length & 0xFF), err, TAG, "write TXPLL failed");
+    ESP_GOTO_ON_ERROR(ch390_io_register_write(emac, CH390_TXPLH, (length >> 8) & 0xFF), err, TAG, "write TXPLH failed");
+
+    /* issue tx polling command */
+    ESP_GOTO_ON_ERROR(ch390_io_register_read(emac, CH390_TCR, &tcr), err, TAG, "read TCR failed");
+    ESP_GOTO_ON_ERROR(ch390_io_register_write(emac, CH390_TCR, tcr | TCR_TXREQ), err, TAG, "write TCR failed");
+    return ESP_OK;
+err:
+    return ret;
+}
+
+static esp_err_t emac_ch390_receive(esp_eth_mac_t *mac, uint8_t *buf, uint32_t *length)
+{
+    esp_err_t ret = ESP_OK;
+    emac_ch390_t *emac = __containerof(mac, emac_ch390_t, parent);
+
+    uint8_t ready;
+    /* dummy read, get the most updated data */
+    ESP_GOTO_ON_ERROR(ch390_io_register_read(emac, CH390_MRCMDX, &ready), err, TAG, "read MRCMDX failed");
+    ESP_GOTO_ON_ERROR(ch390_io_register_read(emac, CH390_MRCMDX, &ready), err, TAG, "read MRCMDX failed");
+
+    // if ready != 1 or 0 reset device
+    if (ready & CH390_PKT_ERR) {
+        emac_ch390_stop(mac);
+        esp_rom_delay_us(1000);
+        emac_ch390_start(mac);
```

Comment:
The section on packet reception is indeed very vague. I've appended the following note at the beginning of the document to help users understand:
```c
/** @note -----------------------  RX Pack Structure ---------------------------
 * |              4 Bytes Frame Head               | Data Area(pass to lwip) |
 * |  Head |  Status  | Length(Low) | Length(High) |        ........         |
 *     |        |
 *     |        |       
 *     |        |       Should be the value of RSR( @ref CH390_RSR). We use @ref RSR_ERR_MASK to determine
 *     |        |       whether the pack has error.
 *     |        |-------------------------------------------------------------------------------------
 *     |
 *     |                Depends on RCSEN bit( @ref RCSCSR_RCSEN) of RCSCSR( @ref CH390_RCSCSR)
 *     |                - RCSEN = 0, the Head shoud always be 0x01
 *     |                - RCSEN = 1, bit 7:2 of the Head is the same as that of RCSCSR;
 *     |                This will affect the determination of the validity of the packet. Therefore, 
 *     |                we provide discriminant masks for both cases.
 *     |                - RCSEN = 0 ---> @ref CH390_PKT_ERR
 *     |                - RCSEN = 1 ---> @ref CH390_PKT_ERR_WITH_RCSEN
 *     |----------------------------------------------------------------------------------------------
*/

```